src/webgpu/util/math.ts

```diff
@@ -875,3 +875,35 @@ export function cartesianProduct<T>(...inputs: T[][]): T[][] {
 
   return result;
 }
+
+/** @returns all of the permutations of an array
+ *
+ * Recursively calculates all of the permutations, does not cull duplicate
+ * entries.
+ *
+ * @param input the array to get permutations of
+ */
+export function calculatePermutations<T>(input: T[]): T[][] {
+  if (input.length === 0) {
+    return [];
+  }
+
+  if (input.length === 1) {
+    return [input];
+  }
+
+  if (input.length === 2) {
+    return [input, [input[1], input[0]]];
+  }
+
+  const result: T[][] = [];
+  input.forEach((head, idx) => {
```

Comment:
Nifty!

Reply:
Thanks, I probably could re-organize this to be iterative instead of recursive, but I am leaning towards doing the thing I can prove is correct/implement easily, and iterate if it is too hot, since it is only being used for 3 and 4 elements arrays.